docs: rework mdbook (#1036)
coreos/clair: Documentation/howto/deployment.md

> # Deploying ClairV4

Reviewer: ```suggestion
# Deploying Clair
```

Reviewer: In general in this doc, I'm removing "node" (easily confused with k8s nodes), referring to Clair's logical services as "processes" and always adding the "OS" qualifier when talking about a linux process.

> ClairV4 was designed with flexible deployment architectures in mind. The operator is free to choose a deployment model which scales to their use cases. 

Reviewer: ```suggestion
Clair v4 was designed with flexible deployment architectures in mind. The operator is free to choose a deployment model which scales to their use cases.
```

## Configuration

> Before jumping directly into the models its important to note that ClairV4 is capable of using a single configuration file across all node types. This design decision makes it very easy to deploy on systems like Kubernetes and Openshift.

Reviewer: ```suggestion
Before jumping directly into the models its important to note that Clair is designed to use a single configuration file across all node types. This design decision makes it very easy to deploy on systems like Kubernetes and OpenShift.
```

OpenShift is inter-capped in its documentation.

See [Config Reference](../reference/config.md)

## Combined Deployment

> In a combined deployment all node types run on in a single Clair process. This is by far the easiest deployment model to configure as it involves the least moving parts. 

Reviewer: ```suggestion
In a combined deployment, all the Clair processes run in a single OS process. This is by far the easiest deployment model to configure as it involves the least moving parts. 
```

> A load balancer is still recommended if you plan on performing TLS termination. Typically this will be a Openshift route or a Kubernetes ingress.

Reviewer: ```suggestion
A load balancer is still recommended if you plan on performing TLS termination. Typically this will be a OpenShift route or a Kubernetes ingress.
```

![combo mode single db deployment diagran](./clairv4_combo_single_db.png)

> In the above diagram Clair is running in combo mode and talking to a single database. To configure this model you will provide all node types the same database and start clair in **combo** mode.

Reviewer: ```suggestion
In the above diagram, Clair is running in combo mode and talking to a single database. To configure this model you will provide all node types the same database and start Clair in **combo** mode.
```

```
...
indexer:
    connstring: "host=clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
matcher:
    connstring: "host=cher-clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
    ...
notifier:
    connstring: "host=clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
    ...
```

> In this mode any configuration informing Clair how to talk to other nodes is ignored, it is not needed as all api communication is done in process.

Reviewer: ```suggestion
In this mode, any configuration informing Clair how to talk to other nodes is ignored, it is not needed as all intra-process communication is done directly.
```

> For added flexibility its also supported to split the databases while in combo mode.

Reviewer: ```suggestion
For added flexibility, it's also supported to split the databases while in combo mode.
```

![combo mode multiple db deployment diagran](./clairv4_combo_multi_db.png)

> In the above diagram Clair is running in combo mode but database load is split between multiple databases. Since Clair is conceptually a set of micro-services, nodes communicate over API and do not share database tables.

Reviewer: ```suggestion
In the above diagram, Clair is running in combo mode but database load is split between multiple databases. Since Clair is conceptually a set of micro-services, its processes do not share database tables even when combined into the same OS process.
```

> To configure this model you would provide each node type it's own "connstring" in the configuration. 

Reviewer: ```suggestion
To configure this model, you would provide each process its own "connstring" in the configuration. 
```

```
...
indexer:
    connstring: "host=indexer-clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
matcher:
    connstring: "host=matcher-clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
    ...
notifier:
    connstring: "host=notifier-clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
    ...
```

## Distributed Deployment

> If your application needs to asymetrically scale or you expect high load you may want to consider a distributed deployment.

Reviewer: ```suggestion
If your application needs to asymmetrically scale or you expect high load you may want to consider a distributed deployment.
```

> In a distributed deployment each Clair node type runs in its own process. Typically this will be a Kubernetes\Openshift deployment or your public cloud of choice's auto scaling mechanism.

Reviewer: ```suggestion
In a distributed deployment, each Clair process runs in its own OS process. Typically this will be a Kubernetes or OpenShift Pod.
```

> A load balancer **must** be setup in this deployment model. The load balancer will route traffic between Clair nodes along with routing API requests via [path based routing](https://devcentral.f5.com/s/articles/the-three-http-routing-patterns-you-should-know-30764) to the correct services. In a Kubernetes or Openshift deployment this is usually trivialized by the Service and Routes abstractions. If you will be deploying on bare metal you will need to configure a load balancer appropriately. 

Reviewer: ```suggestion
A load balancer **must** be setup in this deployment model. The load balancer will route traffic between Clair nodes along with routing API requests via [path based routing](https://devcentral.f5.com/s/articles/the-three-http-routing-patterns-you-should-know-30764) to the correct services. In a Kubernetes or OpenShift deployment this is usually handled with the Service and Routes abstractions. If deploying on bare metal, a load balancer will need to be configured appropriately. 
```

![distributed mode multiple db deployment diagran](./clairv4_distributed_multi_db.png)

> In the above diagram a load balancer is configured to route traffic coming from the client to the correct service. This routing is path based routing and requires a layer 7 load balancer. Traefik, Nginx, and HAProxy are all capable of this. As mentioned above this funtionality is native to Openshift and Kubernetes.

Reviewer: ```suggestion
In the above diagram, a load balancer is configured to route traffic coming from the client to the correct service. This routing is path based routing and requires a layer 7 load balancer. Traefik, Nginx, and HAProxy are all capable of this. As mentioned above, this functionality is native to OpenShift and Kubernetes.
```

> In this configuration you'd supply each node type with their own database conn string and inform each node type how to contact their dependent services. Each node will need to have it's mode CLI flag or ENV variable set to the appropriate node type. 

Reviewer: ```suggestion
In this configuration, you'd supply each process with database connection strings and addresses for their dependent services. Each OS process will need to have its "mode" CLI flag or environment variable set to the appropriate value. 
```

See [Config Reference](../reference/config.md)

```
...
indexer:
    connstring: "host=indexer-clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
matcher:
    connstring: "host=matcher-clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
    indexer_addr: "indexer-service"
    ...
notifier:
    connstring: "host=notifier-clairdb user=pqgotest dbname=pqgotest sslmode=verify-full"
    indexer_addr: "indexer-service"
    matcher_addr: "matcher-service"
    ...
```

> Keep in mind you do not need a config file per node type. You may provide each node the same config and each node will configure itself correctly, only using the values necessary for their configured mode.

Reviewer: ```suggestion
Keep in mind a config file per process is not need. Processes only use the values necessary for their configured mode.
```

## TLS Termination

> Currently ClairV4 punts TLS termination to the load balancing infrastucture. This is a design choice due to the ubiquity of Kubernetes and Openshift infrastructure providing this facility. ClairV4 requires a load balancing solution to terminate TLS.

Reviewer: ```suggestion
Currently Clair offloads TLS termination to the load balancing infrastructure. This design choice is due to the ubiquity of Kubernetes and OpenShift infrastructure already providing this facility.
```

## More On Path Routing

If you are considering a distributed deployment you will need more details on [path based routing](https://devcentral.f5.com/s/articles/the-three-http-routing-patterns-you-should-know-30764). 

Learn how to grab our OpenAPI spec [here](./api.mw) and either start up a local dev instance of the swagger editor or load the spec file into the [online editor](https://petstore.swagger.io/#/)

You will notice particular API paths are grouped by the services which implement them. This is your guide to configure your layer 7 load balancer correctly. 

> When the loadbalancer encounters a particular path prefix it must send those request to the correct set of ClairV4 nodes. 

Reviewer: ```suggestion
When the load balancer encounters a particular path prefix it must send those request to the correct set of Clair nodes. 
```

> For example this is how we configure Traefik in our local development environment.

Reviewer: ```suggestion
For example, this is how we configure Traefik in our local development environment:
```
